### PR TITLE
Lms 2261 handle form elements in pagination

### PR DESCRIPTION
**Pull request by lzabo**
**Description**
- Handle pagination with form elements.
- Add require login.
- Set access to the report page
- Define error types as a constants.

**Related Issues**
- LMS-1625, LMS-2285, LMS-2261, LMS-2262

**PR Type**
Fix issues

### CODE REVIEW

File: report.php [hunk 1]
```diff
@@ -75,43 +123,55 @@
...
-    if (!empty($fromform->userfullname)) {
+    // Set onpage because moodle_url function is not handling arrays.
+    $formelements = clone $fromform;
+
+    unset($formelements->resend, $formelements->submitbutton);
+
+    $urlparams['onpage'] = json_encode($formelements);
+}
+
+echo "<br>here: ";
```
Comment (gordonmacqueen-lp) [inline report.php]: remove this line

File: report.php [hunk 2]
```diff
@@ -75,43 +123,55 @@
...
+if (isset($formelements)) {
+    echo "we have fromform!!!!";
```
Comment (gordonmacqueen-lp) [inline report.php]: remove this line

File: report.php [hunk 3]
```diff
@@ -214,14 +274,13 @@
     }
 }
 
+// Set pagination url.
+$paginationurl =new moodle_url('/admin/tool/log/store/xapi/report.php', $urlparams);
```
Comment (gordonmacqueen-lp) [inline report.php]: space needed: `=new`

**Review by gordonmacqueen-lp — CHANGES_REQUESTED** [review]

**Review by mlynn-lp — APPROVED** [review]

File: lib.php [hunk 4]
```diff
@@ -28,6 +28,12 @@
 define('XAPI_REPORT_SOURCE_FAILED', 'logstore_xapi_failed_log');
 define('XAPI_REPORT_SOURCE_HISTORICAL', 'logstore_standard_log');
 
+// Error types
+define('XAPI_REPORT_ERRORTYPE_NETWORKERROR', 101);
+define('XAPI_REPORT_ERRORTYPE_RECIPEERROR', 400);
+define('XAPI_REPORT_ERRORTYPE_AUTHERROR', 401);
+define('XAPI_REPORT_ERRORTYPE_LRSERROR', 500);
```
Comment (stephen-oh) [inline lib.php]: It's good to add constants for these but as discussed on the call this morning, these aren't correct

File: report.php [hunk 5]
```diff
@@ -21,40 +21,88 @@
...
 $perpage      = optional_param('perpage', XAPI_REPORT_PERPAGE_DEFAULT, PARAM_INT);
+$onpage       = optional_param('onpage', XAPI_REPORT_ONPAGE_DEFAULT, PARAM_TEXT);
 
-$baseurl = new moodle_url('/admin/tool/log/store/xapi/report.php', array('id' => $id, 'page' => $page, 'perpage' => $perpage));
+// Set pagination url's parameter.
+$urlparams = array(
+    'id' => $id,
+    'page' => $page,
+    'perpage' => $perpage,
+    'onpage' => $onpage
```
Comment (stephen-oh) [inline report.php:60]: Do we really need 2 params for which page the user is on?

File: lib.php (hunk 4 above)
Comment (mlynn-lp) [reply lib.php]: We'd need to update the corresponding error code when the failed to transform error occurs.

Comment (stephen-oh) [reply lib.php]: @lzabo can we remove the changes to this function from this PR? I have made changes to this in another PR and we can address this there

File: report.php (hunk 5 above)
Comment (lzabo) [reply report.php:60]: Yes, we need those.
The perpage shows how many element is shown per page.
The page shows which page is shown.
The onpage is a helper to bring the filters' settings from first page to other pages.

**Review by gordonmacqueen-lp — APPROVED** [review]

File: report.php [hunk 6]
```diff
@@ -91,6 +91,8 @@
             $defaults[$element] = $formelements->$element;
         }
     }
+
+    $default['resend'] = XAPI_REPORT_RESEND_FALSE;
```
Comment (gordonmacqueen-lp) [inline report.php]: Can you remove this line so I can merge the PR?

I tested it locally and I'm still getting the same behaviour as reported on the ticket.

Comment (lzabo) [reply report.php]: I have removed this.

**Review by gordonmacqueen-lp — APPROVED** [review]

**Pull request merged**